[CALCITE-6715] Enhance RelFieldTrimmer to trim LogicalCorrelate nodes

kgyrtkirk opened the pull request.

mihaibudiu commented on core/src/test/java/org/apache/calcite/sql2rel/RelFieldTrimmerTest.java:
```diff
@@ -544,4 +548,36 @@ public static Frameworks.ConfigBuilder config() {
+
+    final RelFieldTrimmer fieldTrimmer = new RelFieldTrimmer(null, builder);
+    final RelNode trimmed = fieldTrimmer.trim(root);
+    final String expected = ""
```
a bit difficult to see what the trimmer has done.
perhaps if you have the expected tree before the trim it will be easier?
then you can also assert that they are different

kgyrtkirk replied:
I've updated the testcase to use an `Uncollect` instead - and also added the before-trim plan to the test.

suibianwanwank replied:
Can you add a more generalized example, where both left and right are SCAN and are correlate by some field, and both left and right be trimmed.

kgyrtkirk replied:
I've added another test which removes an unecessarily projected expression from both sides

mihaibudiu commented on core/src/main/java/org/apache/calcite/sql2rel/RelFieldTrimmer.java:
```diff
@@ -1336,6 +1338,132 @@ public TrimResult trimFields(
+    private final RelDataType newCorrelRowType;
+    private final RexBuilder rexBuilder;
+
+    RexCorrelVariableMapShuttle(final CorrelationId correlationId,
```
Even though the class is not public, I think JavaDoc for the parameters would make this easier to understand.

mihaibudiu commented on core/src/main/java/org/apache/calcite/sql2rel/RelFieldTrimmer.java:
```diff
@@ -1336,6 +1338,132 @@ public TrimResult trimFields(
+   * [ 1:0, 2:1, 101:2, 102:3 ] ; sourceCount:200
+   * </pre>
+   */
+  private Mapping makeMapping(List<Mapping> inputMappings) {
```
concatenateMappings is a more suggesting name.

mihaibudiu commented on core/src/main/java/org/apache/calcite/sql2rel/RelFieldTrimmer.java:
```diff
@@ -1336,6 +1338,132 @@ public TrimResult trimFields(
+        correlate
+            .copy(correlate.getTraitSet(),
+                newInputs.get(0),
+                newInputs.get(1).accept(new RexRewritingRelShuttle(rexVisitor)),
```
I find it easier to debug if I avoid complex expressions - I would use a temporary for the result of the visitor.

mihaibudiu replied:
this also assumes implicitly that newInputs.size() == 2; perhaps this is obvious, but then it would not harm to assert that this holds for the correlate

kgyrtkirk replied:
extracted them to be `newLeft` and `newRight` (I was thinking the same earlier ; but I've only extracted the `rexVistior`)

`LogicalCorrelate`'s constructor has only `left` and `right` - but that may change...I've added a check to ensure that this only applies when `correlate.getInputs().size()` is 2

kgyrtkirk commented:
Thank you @mihaibudiu  for taking a look!
I've updated the PR according to your suggestions - kept the new changes as a separate commit for now.
Could you please take another look?

suibianwanwank commented on core/src/test/resources/sql/sub-query.iq:
```diff
@@ -481,7 +481,7 @@ EnumerableCalc(expr#0..2=[{inputs}], proj#0..1=[{exprs}])
     EnumerableValues(tuples=[[{ 1, 2 }]])
     EnumerableAggregate(group=[{0}])
-      EnumerableCalc(expr#0..7=[{inputs}], expr#8=[true], expr#9=[CAST($t7):INTEGER], expr#10=[$cor0], expr#11=[$t10.A], expr#12=[=($t9, $t11)], i=[$t8], $condition=[$t12])
+      EnumerableCalc(expr#0..7=[{inputs}], expr#8=[true], expr#9=[CAST($t7):INTEGER], expr#10=[$cor0], expr#11=[$t10.EXPR$0], expr#12=[=($t9, $t11)], i=[$t8], $condition=[$t12])
```
Why has it changed here, and why hasn't it changed below when the plan below is the same as here?

kgyrtkirk replied:
Actually both of them are misleading; there is no `A` in either plans...due to the fact that `RexInputRef` accesses column by idx not by name - it has no real effect

### for the 1st case the following happens:

<details>
<summary>RelFieldTrimmer match</summary>

matches on the plan:
```
LogicalCorrelate(correlation=[$cor0], joinType=[inner], requiredColumns=[{0}])
  LogicalProject(EXPR$0=[$0], EXPR$1=[$1])
    LogicalValues(tuples=[[{ 1, 2 }]])
  LogicalAggregate(group=[{0}])
    LogicalProject(i=[true])
      LogicalProject(EMPNO=[$0], ENAME=[$1], JOB=[$2], MGR=[$3], HIREDATE=[$4], SAL=[$5], COMM=[$6], DEPTNO=[$7])
        LogicalFilter(condition=[=($8, $cor0.A)])
          LogicalProject(EMPNO=[$0], ENAME=[$1], JOB=[$2], MGR=[$3], HIREDATE=[$4], SAL=[$5], COMM=[$6], DEPTNO=[$7], DEPTNO0=[CAST($7):INTEGER])
            LogicalTableScan(table=[[scott, EMP]])
```
* left is left intact
* the right is trimmed from:
```
LogicalAggregate(group=[{0}])
  LogicalProject(i=[true])
    LogicalProject(EMPNO=[$0], ENAME=[$1], JOB=[$2], MGR=[$3], HIREDATE=[$4], SAL=[$5], COMM=[$6], DEPTNO=[$7])
      LogicalFilter(condition=[=($8, $cor0.A)])
        LogicalProject(EMPNO=[$0], ENAME=[$1], JOB=[$2], MGR=[$3], HIREDATE=[$4], SAL=[$5], COMM=[$6], DEPTNO=[$7], DEPTNO0=[CAST($7):INTEGER])
          LogicalTableScan(table=[[scott, EMP]])

```
to:
```
LogicalAggregate(group=[{0}])
  LogicalProject(i=[true])
    LogicalFilter(condition=[=($1, $cor0.A)])
      LogicalProject(EMPNO=[$0], DEPTNO0=[CAST($7):INTEGER])
        LogicalTableScan(table=[[scott, EMP]])
```
* trim is executed
* the rowType used for left is `RecordType(INTEGER EXPR$0, INTEGER EXPR$1)` (as it has the Project at this point)
* during the rewrite: a `RexFieldAccess` is created to reference column `0`
</details>

<details>
<summary>ProjectRemoveRule fires</summary>

the project of:
```
LogicalProject(EXPR$0=[$0], EXPR$1=[$1])
  LogicalValues(subset=[rel#99:RelSubset#0.NONE.[]], tuples=[[{ 1, 2 }]])
```
Is identified as [isTrivial](https://github.com/apache/calcite/blob/0eb83b1dc6316cba73c2c49e8fbccfa390333fbe/core/src/main/java/org/apache/calcite/rel/rules/ProjectRemoveRule.java#L82)

The project is removed.
</details>


### in the other case:

<details>
<summary>RelFieldTrimmer matches on a different plan</summary>

```
LogicalCorrelate(correlation=[$cor0], joinType=[inner], requiredColumns=[{0}])
  LogicalProject(A=[$0], B=[$1])
    LogicalValues(tuples=[[{ 1, 2 }]])
  LogicalAggregate(group=[{0}])
    LogicalProject(i=[true])
      LogicalProject(EMPNO=[$0], ENAME=[$1], JOB=[$2], MGR=[$3], HIREDATE=[$4], SAL=[$5], COMM=[$6], DEPTNO=[$7])
        LogicalFilter(condition=[=($8, $cor0.A)])
          LogicalProject(EMPNO=[$0], ENAME=[$1], JOB=[$2], MGR=[$3], HIREDATE=[$4], SAL=[$5], COMM=[$6], DEPTNO=[$7], DEPTNO0=[CAST($7):INTEGER])
            LogicalTableScan(table=[[scott, EMP]])
```

* which has `LogicalProject(A=[$0], B=[$1])` ; so the rowtype for left is `RecordType(INTEGER A, INTEGER B)`
* similar transform happens ; but rowType has `A` in it.

</details>
<details>
<summary>ProjectRemoveRule fires</summary>

the project of:
```
LogicalProject(A=[$0], B=[$1])
  LogicalValues(subset=[rel#100:RelSubset#0.NONE.[]], tuples=[[{ 1, 2 }]])
```
Is identified as [isTrivial](https://github.com/apache/calcite/blob/0eb83b1dc6316cba73c2c49e8fbccfa390333fbe/core/src/main/java/org/apache/calcite/rel/rules/ProjectRemoveRule.java#L82)

The project is removed.
</details>

rubenada commented on core/src/main/java/org/apache/calcite/sql2rel/RelFieldTrimmer.java:
```diff
@@ -1336,6 +1338,144 @@ public TrimResult trimFields(
+   * [ 1:0, 2:1, 101:2, 102:3 ] ; sourceCount:200
+   * </pre>
+   */
+  private Mapping concatenateMappings(List<Mapping> inputMappings) {
```
IMO this method might be re-used on other contexts, so it could be moved as a `public static` utility method inside `Mappings` class.

rubenada commented on core/src/main/java/org/apache/calcite/sql2rel/RelFieldTrimmer.java:
```diff
@@ -1336,6 +1338,144 @@ public TrimResult trimFields(
+        Mappings.create(MappingType.INVERSE_SURJECTION, fieldCount, newFieldCount);
+    int offset = 0;
+    int newOffset = 0;
+    for (int i = 0; i < inputMappings.size(); i++) {
```
nit: we could have this for loop simplified as `for (Mapping inputMapping : inputMappings) {...}`

rubenada commented on core/src/main/java/org/apache/calcite/sql2rel/RelFieldTrimmer.java:
```diff
@@ -1336,6 +1338,144 @@ public TrimResult trimFields(
+   */
+  public TrimResult trimFields(LogicalCorrelate correlate,
+      ImmutableBitSet fieldsUsed, Set<RelDataTypeField> extraFields) {
+    if (!extraFields.isEmpty() || correlate.getInputs().size() != 2) {
```
Correlate is a BiRel, so the condition `correlate.getInputs().size() != 2` seems unnecessary.

mihaibudiu commented:
Lint didn't like your commit message - too bad, since it will go away anyway

mihaibudiu commented:
This is better, but @rubenada made some very good suggestions, so I will review once you have addressed these too.

kgyrtkirk commented:
I think I've addressed all comments so-far in [this commit](https://github.com/apache/calcite/pull/4077/commits/18d03e36d904441b0c91931c3f47f2d66c786a93)

Lint is kinda angry because I've multiple commits on the branch and not squashed the changes to be a single commit...would you prefer to squash them? 

...maybe I should use conforming format for all commits next time; that could keep the `LintTest` happy :)

kgyrtkirk commented:
I've force pushed the branch to get a proper CI run; I'll add `LintTest` compliant commits from now on :)

mihaibudiu commented:
If there are no other comments this looks good to merge. Thank you.

kgyrtkirk commented:
Thank you @mihaibudiu , @rubenada and @suibianwanwank for reviewing the changes!
